wgpu/src/dispatch.rs

```diff
@@ -0,0 +1,573 @@
+    fn end_pipeline_statistics_query(&mut self);
+
+    fn execute_bundles(&mut self, render_bundles: &mut dyn Iterator<Item = &AbstractRenderBundle>);
+
+    fn end(&mut self);
+}
+
+pub trait RenderBundleEncoderInterface: Any {
+    fn set_pipeline(&mut self, pipeline: &AbstractRenderPipeline);
+    fn set_bind_group(
+        &mut self,
+        index: u32,
+        bind_group: Option<&AbstractBindGroup>,
+        offsets: &[crate::DynamicOffset],
+    );
+    fn set_index_buffer(
+        &mut self,
+        buffer: &AbstractBuffer,
+        index_format: crate::IndexFormat,
+        offset: crate::BufferAddress,
+        size: Option<crate::BufferSize>,
+    );
+    fn set_vertex_buffer(
+        &mut self,
+        slot: u32,
+        buffer: &AbstractBuffer,
+        offset: crate::BufferAddress,
+        size: Option<crate::BufferSize>,
+    );
+    fn set_push_constants(&mut self, stages: crate::ShaderStages, offset: u32, data: &[u8]);
+
+    fn draw(&mut self, vertices: Range<u32>, instances: Range<u32>);
+    fn draw_indexed(&mut self, indices: Range<u32>, base_vertex: i32, instances: Range<u32>);
+    fn draw_indirect(
+        &mut self,
+        indirect_buffer: &AbstractBuffer,
+        indirect_offset: crate::BufferAddress,
+    );
+    fn draw_indexed_indirect(
+        &mut self,
+        indirect_buffer: &AbstractBuffer,
+        indirect_offset: crate::BufferAddress,
+    );
+
+    fn finish(&mut self, desc: &crate::RenderBundleDescriptor<'_>) -> AbstractRenderBundle;
+}
+
+pub trait CommandBufferInterface: Any + Drop + WasmNotSendSync {}
+pub trait RenderBundleInterface: Any + Drop + WasmNotSendSync {}
+
+pub trait SurfaceInterface: Any + Drop + WasmNotSendSync {
+    fn get_capabilities(&self, adapter: &AbstractAdapter) -> wgt::SurfaceCapabilities;
+
+    fn configure(&self, device: &AbstractDevice, config: &crate::SurfaceConfiguration);
+    fn get_current_texture(
+        &self,
+    ) -> (
+        Option<AbstractTexture>,
+        crate::SurfaceStatus,
+        AbstractSurfaceOutputDetail,
+    );
+    fn present(&self, detail: &AbstractSurfaceOutputDetail);
+    fn texture_discard(&self, detail: &AbstractSurfaceOutputDetail);
+}
+
+pub trait SurfaceOutputDetailInterface: Any + WasmNotSendSync {}
+
+pub trait QueueWriteBufferInterface: Any + WasmNotSendSync + Debug {
+    fn slice(&self) -> &[u8];
+
+    fn slice_mut(&mut self) -> &mut [u8];
+}
+
+pub trait BufferMappedRangeInterface: Any + WasmNotSendSync + Debug {
+    fn slice(&self) -> &[u8];
+    fn slice_mut(&mut self) -> &mut [u8];
+}
+
+macro_rules! interface_type {
+    ($name:ident<$core:ty, $webgpu:ty>: $trait:ident) => {
+        pub enum $name {
+            #[cfg(wgpu_core)]
+            Core($core),
+            #[cfg(webgpu)]
+            WebGPU($webgpu),
+        }
+
+        impl $name {
+            #[cfg(wgpu_core)]
+            #[inline]
+            #[allow(unused)]
+            pub fn as_core(&self) -> &$core {
```

Comment:
`Result` would be nicer imho

Reply:
We'll get there, no idea what the current behavior is if you mix context types, at least now it'll panic in a nice well defined place.